crmprtd/insert.py

```diff
@@ -8,11 +8,217 @@
 speed and reliability. This phase is common to all networks.
 """
 
+from math import floor
+from sqlalchemy import and_
+import logging
+import time
+from sqlalchemy.exc import IntegrityError
+
+from crmprtd.db_exceptions import UniquenessError, InsertionError
+from pycds import Obs
+
+
+log = logging.getLogger('crmprtd.insert')
+
+
+class DBMetrics(object):
+    '''Keep track of database metrics during the insertion process.
+    '''
+
+    def __init__(self):
+        self.successes = 0
+        self.failures = 0
+
+    def clear(self):
+        self.successes = 0
+        self.failures = 0
+
+
+class Timer(object):
+    def __enter__(self):
+        self.start = time.time()
+        return self
+
+    def __exit__(self, *args):
+        self.end = time.time()
+        self.run_time = self.end - self.start
+
+
+def chunks(list, chunk_size):
+    for i in range(0, len(list), chunk_size):
+        yield list[i:i+chunk_size]
+
+
+def get_sample_indices(num_obs, sample_size):
+    if sample_size > num_obs:
+        sample_size = num_obs
+
+    index_buffer = floor(num_obs/sample_size)
+
+    sample_indices = [0]
+    index = 0
+    for i in range(sample_size - 1):
+        index += index_buffer
+        sample_indices.append(index)
+
+    return sample_indices
+
+
+def is_unique(sesh, history_id, vars_id, time):
+    q = sesh.query(Obs.id).filter(
+        and_(Obs.history_id == history_id, Obs.vars_id == vars_id,
+             Obs.time == time))
+    if q.count() > 0:
+        return False
+
+    return True
+
+
+def has_unique_obs(sesh, observations, sample_size):
+    sample_obs = []
+    sample_indices = get_sample_indices(len(observations), sample_size)
+
+    for index in sample_indices:
+        sample_obs.append(observations[index])
+
+    already_exists = 0
+    for sample in sample_obs:
+        if not is_unique(sesh, sample.history_id, sample.vars_id, sample.time):
+            already_exists += 1
+
+    if already_exists == sample_size:
+        return False
+
+    return True
+
+
+def single_insert_obs(sesh, o, dbm):
+    # Check to see if this entry will be unique
+    try:
+        if not is_unique(sesh, o.history_id, o.vars_id, o.time):
+            log.warning("Failure, observation already exists.")
+            dbm.failures += 1
+            raise UniquenessError(o)
+    except UniquenessError as e:
+        log.warning("Failure, observation already exists.", extra={'e': e})
+        dbm.failures += 1
+        raise e
+    except Exception as e:
+        log.warning("Failure, an error occured.", extra={'e': e})
+        dbm.failures += 1
+        raise InsertionError(obs_time=o.time, datum=o.datum,
+                             vars_id=o.vars_id, hid=o.history_id, e=e)
+
+    # value does not exist in obs_raw, continue with insertion
+    try:
+        sesh.add(o)
+        sesh.commit()
+        log.debug("Successfully inserted observation")
+        dbm.successes += 1
+        return 1
+    except Exception as e:
```

Comment:
This *is* appropriate exception handling, because an error isn't something that we "expect" I.e. it's *exceptional* behaviour.